nixos/modules/services/monitoring/prometheus/exporters/restic.nix

```diff
@@ -0,0 +1,114 @@
+        repository to monitor.
+      '';
+      example = "sftp:backup@192.168.1.100:/backups/example";
+    };
+
+    passwordFile = mkOption {
+      type = types.str;
+      description = lib.mdDoc ''
+        Read the repository password from a file.
+      '';
+      example = "/etc/nixos/restic-password";
+    };
+
+    environmentFile = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      description = lib.mdDoc ''
+        file containing the credentials to access the repository, in the
+        format of an EnvironmentFile as described by systemd.exec(5)
+      '';
+    };
+
+    rcloneOptions = mkOption {
+      type = with types; nullOr (attrsOf (oneOf [ str bool ]));
+      default = null;
+      description = lib.mdDoc ''
+        Options to pass to rclone to control its behavior.
+        See <https://rclone.org/docs/#options> for
+        available options. When specifying option names, strip the
+        leading `--`. To set a flag such as
+        `--drive-use-trash`, which does not take a value,
+        set the value to the Boolean `true`.
+      '';
+    };
+
+    rcloneConfig = mkOption {
+      type = with types; nullOr (attrsOf (oneOf [ str bool ]));
+      default = null;
+      description = lib.mdDoc ''
+        Configuration for the rclone remote being used for backup.
+        See the remote's specific options under rclone's docs at
+        <https://rclone.org/docs/>. When specifying
+        option names, use the "config" name specified in the docs.
+        For example, to set `--b2-hard-delete` for a B2
+        remote, use `hard_delete = true` in the
+        attribute set.
+        Warning: Secrets set in here will be world-readable in the Nix
+        store! Consider using the `rcloneConfigFile`
+        option instead to specify secret values separately. Note that
+        options set here will override those set in the config file.
+      '';
+    };
+
+    rcloneConfigFile = mkOption {
+      type = with types; nullOr path;
+      default = null;
+      description = lib.mdDoc ''
+        Path to the file containing rclone configuration. This file
+        must contain configuration for the remote specified in this backup
+        set and also must be readable by root. Options set in
+        `rcloneConfig` will override those set in this
+        file.
+      '';
+    };
+  };
+
+  serviceOpts = {
+    serviceConfig = {
+      ExecStart = ''
+        ${pkgs.prometheus-restic-exporter}/bin/restic-exporter.py \
+          ${concatStringsSep " \\\n  " cfg.extraFlags}
+      '';
+
+    } // optionalAttrs (cfg.environmentFile != null) {
+      EnvironmentFile = cfg.environmentFile;
+    };
+    environment =
+      let
+        rcloneRemoteName = builtins.elemAt (splitString ":" cfg.repository) 1;
+        rcloneAttrToOpt = v: "RCLONE_" + toUpper (builtins.replaceStrings [ "-" ] [ "_" ] v);
+        rcloneAttrToConf = v: "RCLONE_CONFIG_" + toUpper (rcloneRemoteName + "_" + v);
```

Comment:
This looks like it's trying to do some magic to preserve UX, which is usually a good thing, but this also seems brittle. It's going to break on non-sftp remotes like `rest:https://host:8000/ init`, both because of extra colons and `//`.

Maybe it would be better to avoid implementing a makeshift URI parser, and just let the user specify the remote manually instead?

Reply:
This is honestly just copy-pasted code, and as I don't use rclone, I wouldn't even be opposed to removing the corresponding options entirely.

Reply:
For now, I would not change it, so that Users can copy their preexisting Backup settings, without needing to change anything.

Reply:
How about this: we create a `remoteName` option under `extraOpts` and let it default to `builtins.elemAt (splitString ":" cfg.repository) 1`, so that people at least can change it out if it was to break?

Reply:
I think part of the confusion here is when the `rclone` settings are relevant.

Restic uses rcloe only if you provide it with a Repo URL in the following Format `rclone:<remote>:<path>`, so `rcloneRemoteName` would only extract `<remote>`.

For all other Backends (Rest `rest:http://host:8000/`, S3 `s3:s3.amazonaws.com/bucket_name`, ...) the rclone settings are ignored

Reply:
Technically we could do this to not even generate malformed config Files, but I don't think the effort is worth it.
```diff
-       // (mapAttrs'
+       // optionalAttrs ((builtins.elemAt (splitString ":" cfg.repository) 0) == "rclone") 
+       ((mapAttrs'
         (name: value:
           nameValuePair (rcloneAttrToOpt name) (toRcloneVal value)
         )
         cfg.rcloneOptions)
       // optionalAttrs (cfg.rcloneConfigFile != null) {
         RCLONE_CONFIG = cfg.rcloneConfigFile;
       }
       // (mapAttrs'
         (name: value:
           nameValuePair (rcloneAttrToConf name) (toRcloneVal value)
         )
-         cfg.rcloneConfig);
+         cfg.rcloneConfig));
```
(Untested)

Reply:
I do think something should be done though. `builtins.elem [ "x" ] 1` throws an error, and `/data` (which after `splitString 1` results in `[ "/data" ]`) should be a valid configuration [according to the docs](https://github.com/ngosang/restic-exporter#configuration). The current code would prevent people from setting such a configuration.

Maybe we could set a bogus value to be ignored by the program, in case the repo name is not prefixed by rclone?

```nix
rcloneRemoteName = if lib.hasPrefix "rclone" cfg.repository
  then builtins.elemAt (splitString ":" cfg.repository) 1
  else "IGNORE";
```

Reply:
Yes `/data` is a valid syntax, but that also doesn't go through the rclone backend.
And due to nix being lazy, the error would only be thrown when you set useless rclone Settings.

Reply:
> [...] the error would only be thrown when you set useless rclone Settings.

Right, that makes sense. Then I won't block this anymore. Feel free to mark as resolved.